server/utils/cloudFunctions/functions/test/countAdvertisersTests/countAdvertisersFunctions.test.js

```diff
@@ -0,0 +1,400 @@
+ *   for testing simplicity.
+ *   - See: https://firebase.google.com/docs/functions/get-started
+ * - Test cleanup is performed by an `after` function, that wipes 
+ *   all touched collections. 
+ *   - Firestore doesn't have a default method to delete collections
+ *     so a method has been taken from firestore docs and is in the `test`
+ *     directory as `deleteCollection.js`.   
+ * Author: Robert Marcus
+ * Date: July 13, 2020
+ */
+
+// Import the testing environment configuration
+const { assert,  
+        chai,
+        firestoreWrap, 
+        snapFromJson,
+        DEV_ADS_PATH } = require('../testConfig');
+
+const { deleteCollection } = require('../deleteCollection');
+
+const { DB } = require('../../firebaseConfig');
+
+const { DEV_AGGREGATES_COLLECTION } = 
+    require('../../countAdvertisersCloudFunctions/countAdvertisersConfig');
+
+// Import the cloud functions.
+const { devCountAdvertisersOnCreate, devCountAdvertisersOnDelete } = 
+    require('../../countAdvertisersCloudFunctions/devCountAdvertisersFunctions');
+
+// Save some space for test explanation in the test description
+const SHOULD_INCREMENT = "should increment advertisers ad count "; 
+
+describe('Count advertisers cloud functions', () => {
+  // After every 'describe' block, reset the test environments. 
+  // Since these are live dev db's, this process can be very flaky.
+  // In general, don't expect an 'advertiser' will be reset from 
+  // test to test. 
+  after(() => {
+    const collections = ['dev_ads',
+                         'dev_aggregates/2018/advertisers',
+                         'dev_aggregates/2019/advertisers',
+                         'dev_aggregates/2020/advertisers']
+    collections.forEach(collection => deleteCollection(DB, collection));
+  });
+
+  describe("test_countAdvertisersOnCreate", () => {
+    it(SHOULD_INCREMENT + 'in the correct year', () => {
+      const snap = snapFromJson({advertiser: "adv_a", startDate: "2019-10-15"}, 
+                                DEV_ADS_PATH);
+
+      const wrapper = firestoreWrap(devCountAdvertisersOnCreate);
+      
+      return wrapper(snap).then(() => {
+        DEV_AGGREGATES_COLLECTION
+            .doc("2019")
+            .collection("advertisers")
+            .doc("adv_a").get().then(function(doc) {
+              return assert.equal(doc.data().numberOfAds, 1);
+            }).catch(err => console.log(err));
+
+        return;
+      });
+    });
+
+    it(SHOULD_INCREMENT + 'for 2018, 2019, 2020', () => {
+      const advertiser = "adv_b"; 
+      const snapOne = 
+          snapFromJson({advertiser, startDate: "2018-10-15"}, DEV_ADS_PATH);
+      const snapTwo = 
+          snapFromJson({advertiser, startDate: "2019-10-15"}, DEV_ADS_PATH);
+      const snapThree = 
+          snapFromJson({advertiser, startDate: "2020-10-15"}, DEV_ADS_PATH);
+
+      const wrapper = firestoreWrap(devCountAdvertisersOnCreate);
+
+      return wrapper(snapOne).then(() => {
```

Comment:
You should be able to combine all of these in one statement without this awkward nesting.

Reply:
Done where possible with `Promise.allSettled`.